src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -3,203 +3,211 @@ namespace NServiceBus
                     }
+                }
 
-                    data.AddOrUpdate(sagaData.Id,
-                        id => new VersionedSagaEntity(sagaData, lockenTokenKey),
-                        (id, original) => new VersionedSagaEntity(sagaData, lockenTokenKey, original)); // we can never end up here.
+                var entry = new Entry(sagaData, correlationId);
+                if (_sagas.TryAdd(sagaData.Id, entry) == false)
+                {
+                    throw new Exception("FATAL: this should never happened as saga id should be unique");
                 }
+
+                context.Set(sagaData.Id.ToString(), entry);
             });
+
             return TaskEx.CompletedTask;
         }
 
         public Task Update(IContainSagaData sagaData, SynchronizedStorageSession session, ContextBag context)
         {
-            var inMemSession = (InMemorySynchronizedStorageSession) session;
-            inMemSession.Enlist(() =>
+            ((InMemorySynchronizedStorageSession)session).Enlist(() =>
             {
-                data.AddOrUpdate(sagaData.Id,
-                    id => new VersionedSagaEntity(sagaData, $"{sagaData.GetType().FullName}.None.None"), // we can never end up here.
-                    (id, original) => new VersionedSagaEntity(sagaData, original.LockTokenKey, original));
-            });
-            return TaskEx.CompletedTask;
-        }
+                var entry = GetEntry(sagaData, context);
 
-        void ValidateUniqueProperties(SagaCorrelationProperty correlationProperty, IContainSagaData saga)
-        {
-            var sagaType = saga.GetType();
-            var existingSagas = new List<VersionedSagaEntity>();
-            // ReSharper disable once LoopCanBeConvertedToQuery
-            foreach (var s in data)
-            {
-                if (s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id))
+                if (_sagas.TryUpdate(sagaData.Id, entry.UpdateTo(sagaData), entry) == false)
                 {
-                    existingSagas.Add(s.Value);
+                    throw new Exception($"InMemorySagaPersister concurrency violation: saga entity Id[{ sagaData.Id }] already saved.");
                 }
-            }
-            var uniqueProperty = sagaType.GetProperty(correlationProperty.Name);
+            });
 
-            if (correlationProperty.Value == null)
-            {
-                var message = $"Cannot store saga with id '{saga.Id}' since the unique property '{uniqueProperty.Name}' has a null value.";
-                throw new InvalidOperationException(message);
-            }
+            return TaskEx.CompletedTask;
+        }
 
-            foreach (var storedSaga in existingSagas)
+        static Entry GetEntry(IContainSagaData sagaData, ContextBag context)
+        {
+            Entry entry;
+            if (context.TryGet(sagaData.Id.ToString(), out entry) == false)
```

Comment:
> If you instantiate a saga in a saga finder you break the contract, don't you?

technically yes, not if you create a new instance with the same id and correlation id properties I think.